### PR TITLE
Change GeoIP downloader policy after 30 days of no updates

### PR DESCRIPTION
**Pull request by probakowski**
This PR changes the way `GeoIpDownloader` and `GeoIpProcessor` handle situation when we are unable to update databases for 30 days. In that case:

- `GeoIpDownloader` will delete all chunks from `.geoip_databases` index
- `DatabaseRegistry` will delete all files on ingest nodes
- `GeoIpProcessor` will tag document with `tags: ["_geoip_expired_database"]` field (same way as in Logstash)

This change also fixes bug with that breaks `DatabaseRegistry` and when it tires to download databases after updating timestamp only (`GeoIpDownloader` checks if there are new databases and updates timestamp because local databases are up to date)

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine update branch

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java [hunk 1]
```diff
@@ -166,7 +174,17 @@ void deleteOldChunks(String name, int firstChunk) {
     //visible for testing
     protected void updateTimestamp(String name, Metadata old) {
         logger.info("geoip database [" + name + "] is up to date, updated timestamp");
-        state = state.put(name, new Metadata(System.currentTimeMillis(), old.getFirstChunk(), old.getLastChunk(), old.getMd5()));
+        long timestamp = System.currentTimeMillis();
+        UpdateByQueryRequest request = new UpdateByQueryRequest(DATABASES_INDEX).setRefresh(true);
+        request.setScript(new Script("ctx._source.timestamp=" + timestamp + "L"));
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java:179]: Could we record the timestamp elsewhere instead of adding a field to the chunk documents and updating all these documents? Maybe store it in the cluster state as mapping metadata or custom data in `IndexMetadata`? Under the hood this results in reindexing all the chunk documents, which would be nice if this could be avoided.

File: modules/ingest-geoip/src/test/resources/plugin-security.policy [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+grant {
```
Comment (martijnvg) [inline modules/ingest-geoip/src/test/resources/plugin-security.policy:9]: Why is this needed?

**Review by martijnvg — COMMENTED** [review]
Thanks @probakowski! I left a few questions.

File: modules/ingest-geoip/build.gradle [hunk 3]
```diff
@@ -71,6 +72,9 @@ tasks.named("bundlePlugin").configure {
   }
 }
 
+tasks.named("processTestResources").configure {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/build.gradle]: this was needed to make tests work with painless, I removed it now as we don't use update by query anymore

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java [hunk 4]
```diff
@@ -98,6 +107,9 @@ public IngestDocument execute(IngestDocument ingestDocument) throws IOException
             return ingestDocument;
         } else if (ip == null) {
             throw new IllegalArgumentException("field [" + field + "] is null, cannot extract geoip information.");
+        } if (isValid.get() == false) {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java]: done

File: modules/ingest-geoip/src/test/resources/plugin-security.policy (hunk 2 above)
Comment (probakowski) [reply modules/ingest-geoip/src/test/resources/plugin-security.policy]: this was needed to make tests work with painless, I removed it now as we don't use update by query anymore

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java (hunk 1 above)
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java]: Yeah, I've added additional field in `GeoIpTaskState.Metadata` to use in validity test instead of updating timestamp and it does simplify code greatly. Thanks for suggestion!

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java [hunk 5]
```diff
@@ -63,13 +67,16 @@
         TimeValue.timeValueDays(3), TimeValue.timeValueDays(1), Property.Dynamic, Property.NodeScope);
     public static final Setting<String> ENDPOINT_SETTING = Setting.simpleString("ingest.geoip.downloader.endpoint",
         "https://geoip.elastic.co/v1/database", Property.NodeScope);
+    public static final Setting<TimeValue> DATABASE_VALIDITY = Setting.timeSetting("ingest.geoip.database.validity",
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java]: I removed the constant as it was leftover, but similar setting (`ingest.geoip.database_validity`) is used by tests as you said. It's registered only by test module (similarly to `ingest.geoip.database_path`) so it shouldn't be available in production

[the remark this answers is not in the document]
Comment (martijnvg) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java]: 👍

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java (hunk 1 above)
Comment (martijnvg) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java]: That is great! 👍

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java [hunk 6]
```diff
@@ -202,25 +203,31 @@ void checkDatabases(ClusterState state) {
...
+                    retrieveAndUpdateDatabase(name, metadata);
+                } catch (Exception ex) {
+                    LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("attempt to download database [{}] failed", name), ex);
+                }
+            });
 
         List<String> staleEntries = new ArrayList<>(databases.keySet());
-        staleEntries.removeAll(taskState.getDatabases().keySet());
+        staleEntries.removeAll(taskState.getDatabases().entrySet().stream()
+            .filter(e->e.getValue().isValid(state.getMetadata().settings()))
```
Comment (martijnvg) [inline modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:228]: Maybe I'm not understanding something correctly here. But how are the files for invalid databases removed here?

**Review by martijnvg — COMMENTED** [review]
Looks great! I left a few more comments.

Also I think we should add a section to the documentation that describes when a database can become invalid and what the behaviour is of the geoip processor in that case. Doing this in a followup pr is fine with me too.

File: modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java [hunk 7]
```diff
@@ -78,12 +82,72 @@ protected Settings nodeSettings(int nodeOrdinal, Settings otherSettings) {
...
+            .get();
+        assertTrue(settingsResponse.isAcknowledged());
+    }
+
+    public void testInvalidTimestamp() throws Exception {
+        ClusterUpdateSettingsResponse settingsResponse =
+            client().admin().cluster()
+                .prepareUpdateSettings()
+                .setPersistentSettings(Settings.builder()
+                    .put("ingest.geoip.database_validity", TimeValue.timeValueMillis(1))
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderIT.java]: Good point about verifying that ingestion works prior invalidation. 
Check for no local database files is there already, around line 129

File: modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/GeoIpProcessorTests.java [hunk 8]
```diff
@@ -289,7 +289,7 @@ public void testListFirstOnly() throws Exception {
 
     public void testListFirstOnlyNoMatches() throws Exception {
         GeoIpProcessor processor = new GeoIpProcessor(randomAlphaOfLength(10), null, "source_field",
-            loader("/GeoLite2-City.mmdb"), "target_field", EnumSet.allOf(GeoIpProcessor.Property.class), false, true);
+            loader("/GeoLite2-City.mmdb"), () -> true, "target_field", EnumSet.allOf(GeoIpProcessor.Property.class), false, true);
 
         Map<String, Object> document = new HashMap<>();
         document.put("source_field", Arrays.asList("127.0.0.1", "127.0.0.2"));
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/GeoIpProcessorTests.java:299]: Added

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java [hunk 9]
```diff
@@ -413,7 +427,20 @@ public GeoIpProcessor create(
                     "] doesn't match with expected suffix [" + expectedSuffix + "]";
                 return loader;
             };
-            return new GeoIpProcessor(processorTag, description, ipField, supplier, targetField, properties, ignoreMissing, firstOnly);
+            Supplier<Boolean> isValid = () -> {
+                ClusterState currentState = clusterService.state();
+                if (currentState == null) {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java]: I've changed it

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java [hunk 10]
```diff
@@ -413,7 +427,20 @@ public GeoIpProcessor create(
...
+                    return true;
+                }
+                PersistentTask<?> task = getTaskWithId(currentState, GeoIpDownloader.GEOIP_DOWNLOADER);
+                if (task == null || task.getState() == null) {
+                    return true;
```
[the remark this answers is not in the document]
Comment (probakowski) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpProcessor.java:436]: yes, this is the idea here to let builtin databases to always be valid even if there's no state (i.e. geoip downloader is disabled)

**Comment by probakowski** [comment]
Thanks for the review @martijnvg

> Also I think we should add a section to the documentation that describes when a database can become invalid and what the behaviour is of the geoip processor in that case. Doing this in a followup pr is fine with me too.

I think we will need to add it in PR that will enable geoip downloader again - only then database can get invalid

File: modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java (hunk 6 above)
Comment (martijnvg) [reply modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java:228]: 👍 I see, I missed that.

**Review by martijnvg — APPROVED** [review]
LGTM
